### PR TITLE
Refactor Lexer#next()

### PR DESCRIPTION
**Pull request by tjvr**
This function was getting a bit hard to follow, so I refactored it to make sure that I can understand it. :-)

### CODE REVIEW

File: moo.js [hunk 1]
```diff
@@ -407,10 +407,10 @@
...
   }
 
-  Lexer.prototype._eat = hasSticky ? function(re) { // assume re is /y
-    return re.exec(this.buffer)
-  } : function(re) { // assume re is /g
-    var match = re.exec(this.buffer)
+  Lexer._eat = hasSticky ? function(re, buffer) { // assume re is /y
+    return re.exec(buffer)
+  } : function(re, buffer) { // assume re is /g
+    var match = re.exec(buffer)
```
Comment (tjvr) [inline moo.js]: This should save a lookup on `this`.

File: moo.js [hunk 2]
```diff
@@ -437,66 +433,53 @@
...
-      this.queuedThrow = false
-      if (queuedThrow) {
-        throw new Error(this.formatError(queuedToken, "invalid syntax"))
-      }
-      return queuedToken
+    var index = this.index
+
+    // If a fallback token matched, we don't need to re-run the RegExp
+    if (this.queuedGroup) {
+      var token = this._token(this.queuedGroup, this.queuedText, index)
```
Comment (tjvr) [inline moo.js]: If JS had tuples / multi-return, I'd have written all this code the other way round, e.g. `var [group, text] = this._match(index)`.

File: moo.js [hunk 3]
```diff
@@ -437,66 +433,53 @@
...
+      // Fallback tokens contain the unmatched portion of the buffer
+      return this._token(error, buffer.slice(index, match.index), index)
     }
 
-    if (group.pop) this.popState()
-    else if (group.push) this.pushState(group.push)
-    else if (group.next) this.setState(group.next)
-
-    return fallbackToken || token
+    return this._token(group, text, index)
```
Comment (tjvr) [inline moo.js]: `shouldThrow` and state handling have moved into `this._token()`.

File: moo.js (hunk 1 above)
Comment (nathan) [reply moo.js]: If you're going to do this, why not just do `var eat = …`?
